File: packages/dxtr_pixs_gpu/android/src/main/kotlin/dev/pixelcraft/pixelcraft/NativeGpuCameraLook.kt
package dev.pixelcraft.pixelcraft

/**
 * PF2 native camera-look contract shared by the Android channel/session/renderer.
 *
 * This object is control-plane state only. Camera pixels remain entirely native
 * and Rust remains authoritative for the saved full-resolution result.
 */
internal data class NativeGpuCameraLook(
    val filmProfileId: String = "",
    val filmStrength: Float = 0f,
    val creativeFilterId: String = "",
    val creativeFilterStrength: Float = 0f,
    val brightness: Float = 1f,
    val contrast: Float = 1f,
    val saturation: Float = 1f,
) {
    companion object {
        private val exactCreativeFilters = setOf("grayscale", "invert")
        private val lutCreativeFilters = setOf(
            "vintage",
            "oceanic",
            "lofi",
            "dramatic",
            "golden",
            "pastel_pink",
        )
        val supportedCreativeFilters = exactCreativeFilters + lutCreativeFilters

        fun fromChannel(arguments: Map<*, *>): NativeGpuCameraLook {
            val creative = arguments["creativeFilterId"] as? String ?: ""
            require(creative.isEmpty() || creative in supportedCreativeFilters) {
                "Unsupported camera creative filter: $creative"
            }
            return NativeGpuCameraLook(
                filmProfileId = arguments["filmProfileId"] as? String ?: "",
                filmStrength = number(arguments, "filmStrength", 0f).coerceIn(0f, 1f),
                creativeFilterId = creative,
                creativeFilterStrength =
                    number(arguments, "creativeFilterStrength", 0f).coerceIn(0f, 1f),
                brightness = number(arguments, "brightness", 1f).coerceIn(0f, 2f),
                contrast = number(arguments, "contrast", 1f).coerceIn(0f, 2f),
                saturation = number(arguments, "saturation", 1f).coerceIn(0f, 2f),
            )
        }

        fun creativeLutAssetId(filterId: String): String? = when (filterId) {
            "vintage" -> "creative_vintage"
            "oceanic" -> "creative_oceanic"
            "lofi" -> "creative_lofi"
            "dramatic" -> "creative_dramatic"
            "golden" -> "creative_golden"
            "pastel_pink" -> "creative_pastel_pink"
            else -> null
        }

        private fun number(arguments: Map<*, *>, key: String, fallback: Float): Float =
            (arguments[key] as? Number)?.toFloat() ?: fallback
    }

    val hasFilm: Boolean
        get() = filmProfileId.isNotEmpty() && filmStrength > 0f

    val hasCreative: Boolean
        get() = creativeFilterId.isNotEmpty() && creativeFilterStrength > 0f

    val hasAdjustments: Boolean
        get() = brightness != 1f || contrast != 1f || saturation != 1f

    val isNeutral: Boolean
        get() = !hasFilm && !hasCreative && !hasAdjustments
}

File: packages/dxtr_pixs_gpu/ios/Classes/NativeGpuCameraLook.swift
import Foundation

/// PF2 native camera-look control-plane contract.
///
/// Camera frames never cross Flutter's MethodChannel and this state never
/// becomes final-render authority; Rust remains authoritative for saved pixels.
struct NativeGpuCameraLook {
  static let supportedCreativeFilters: Set<String> = [
    "grayscale",
    "invert",
    "vintage",
    "oceanic",
    "lofi",
    "dramatic",
    "golden",
    "pastel_pink",
  ]

  let filmProfileId: String
  let filmStrength: Float
  let creativeFilterId: String
  let creativeFilterStrength: Float
  let exposure: Float
  let temperature: Float
  let tint: Float
  let brightness: Float
  let contrast: Float
  let saturation: Float
  let vignette: Float

  init(arguments: [String: Any]) throws {
    let creative = arguments["creativeFilterId"] as? String ?? ""
    guard creative.isEmpty || Self.supportedCreativeFilters.contains(creative) else {
      throw Self.error("Unsupported camera creative filter: \(creative)")
    }

    filmProfileId = arguments["filmProfileId"] as? String ?? ""
    filmStrength = Self.clamp(Self.number(arguments, "filmStrength", fallback: 0), 0, 1)
    creativeFilterId = creative
    creativeFilterStrength = Self.clamp(
      Self.number(arguments, "creativeFilterStrength", fallback: 0),
      0,
      1
    )
    exposure = Self.clamp(Self.number(arguments, "exposure", fallback: 0), -2, 2)
    temperature = Self.clamp(Self.number(arguments, "temperature", fallback: 0), -1, 1)
    tint = Self.clamp(Self.number(arguments, "tint", fallback: 0), -1, 1)
    brightness = Self.clamp(Self.number(arguments, "brightness", fallback: 1), 0, 2)
    contrast = Self.clamp(Self.number(arguments, "contrast", fallback: 1), 0, 2)
    saturation = Self.clamp(Self.number(arguments, "saturation", fallback: 1), 0, 2)
    vignette = Self.clamp(Self.number(arguments, "vignette", fallback: 0), -1, 1)
  }

  var hasFilm: Bool {
    !filmProfileId.isEmpty && filmStrength > 0
  }

  var hasCreative: Bool {
    !creativeFilterId.isEmpty && creativeFilterStrength > 0
  }

  var hasAdjustments: Bool {
    exposure != 0 || temperature != 0 || tint != 0 ||
      brightness != 1 || contrast != 1 || saturation != 1 || vignette != 0
  }

  var isNeutral: Bool {
    !hasFilm && !hasCreative && !hasAdjustments
  }

  var creativeLutAssetId: String? {
    switch creativeFilterId {
    case "vintage": return "creative_vintage"
    case "oceanic": return "creative_oceanic"
    case "lofi": return "creative_lofi"
    case "dramatic": return "creative_dramatic"
    case "golden": return "creative_golden"
    case "pastel_pink": return "creative_pastel_pink"
    default: return nil
    }
  }

  private static func number(
    _ arguments: [String: Any],
    _ key: String,
    fallback: Float
  ) -> Float {
    (arguments[key] as? NSNumber)?.floatValue ?? fallback
  }

  private static func clamp(_ value: Float, _ lower: Float, _ upper: Float) -> Float {
    max(lower, min(upper, value))
  }

  private static func error(_ message: String) -> NSError {
    NSError(
      domain: "PixelCraftGpu",
      code: 3201,
      userInfo: [NSLocalizedDescriptionKey: message]
    )
  }
}
